File: drills/python/solutions/nc81_design_twitter__merge_per_user.py
import heapq


class Twitter:
    def __init__(self):
        self.clock = 0
        self.tweets = {}
        self.following = {}

    # Tweets stored per author rather than in one global list, newest last.
    def postTweet(self, userId, tweetId):
        self.tweets.setdefault(userId, []).append((self.clock, tweetId))
        self.clock += 1

    def follow(self, followerId, followeeId):
        self.following.setdefault(followerId, set()).add(followeeId)

    def unfollow(self, followerId, followeeId):
        self.following.setdefault(followerId, set()).discard(followeeId)

    # A k-way merge over the timelines of the people being followed, which only
    # ever needs the ten newest -- so a heap over the tails of the k lists does
    # it without touching every tweet.
    def getNewsFeed(self, userId):
        visible = self.following.get(userId, set()) | {userId}
        candidates = []
        for author in visible:
            candidates.extend(self.tweets.get(author, [])[-10:])
        return [tweetId for _clock, tweetId in heapq.nlargest(10, candidates)]

File: drills/ts/solutions/nc81_design_twitter__merge_per_user.ts
export class Twitter {
  private clock = 0;
  private tweets = new Map<number, [number, number][]>();
  private following = new Map<number, Set<number>>();

  // Tweets stored per author rather than in one global list, newest last.
  postTweet(userId: number, tweetId: number): void {
    if (!this.tweets.has(userId)) this.tweets.set(userId, []);
    this.tweets.get(userId)!.push([this.clock++, tweetId]);
  }

  follow(followerId: number, followeeId: number): void {
    this.followees(followerId).add(followeeId);
  }

  unfollow(followerId: number, followeeId: number): void {
    this.followees(followerId).delete(followeeId);
  }

  // A k-way merge over the timelines of the people being followed, which only
  // ever needs the ten newest -- so a heap over the tails of the k lists does
  // it without touching every tweet.
  getNewsFeed(userId: number): number[] {
    const visible = new Set(this.followees(userId));
    visible.add(userId);

    const candidates: [number, number][] = [];
    for (const author of visible) candidates.push(...(this.tweets.get(author) ?? []).slice(-10));

    return candidates
      .sort((a, b) => b[0] - a[0])
      .slice(0, 10)
      .map(([, tweetId]) => tweetId);
  }

  private followees(userId: number): Set<number> {
    if (!this.following.has(userId)) this.following.set(userId, new Set());
    return this.following.get(userId)!;
  }
}
